internal/biome/replay/record.go

```diff
@@ -0,0 +1,202 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package replay
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/yourbase/yb/internal/biome"
+)
+
+// A Recorder wraps another biome and records its commands in memory.
+// The interactions are saved to the filesystem when it is finished.
+type Recorder struct {
+	biome biome.Biome
+	dir   string
+
+	mu   sync.Mutex
+	data *replayData
+}
+
+// NewRecorder returns a new Recorder biome that wraps the given biome and
+// saves to dir on Close.
+func NewRecorder(dir string, bio biome.Biome) *Recorder {
+	return &Recorder{
+		biome: bio,
+		dir:   dir,
+		data: &replayData{
+			Descriptor:   bio.Describe(),
+			Dirs:         bio.Dirs(),
+			CleanedPaths: make(map[string]string),
+			AbsPaths:     make(map[string]bool),
+		},
+	}
+}
+
+// Close writes the recorded invocations to the directory passed into NewRecorder.
+func (rec *Recorder) Close() error {
+	rec.mu.Lock()
+	fname := recordFilename(rec.data.Descriptor)
+	jsonData, err := json.MarshalIndent(rec.data, "", "\t")
+	rec.mu.Unlock()
+	if err != nil {
+		return fmt.Errorf("close recorder: %w", err)
+	}
+	if err := os.MkdirAll(rec.dir, 0777); err != nil {
+		return fmt.Errorf("close recorder: %w", err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(rec.dir, fname), jsonData, 0666); err != nil {
+		return fmt.Errorf("close recorder: %w", err)
+	}
+	return nil
+}
+
+func recordFilename(desc *biome.Descriptor) string {
+	return desc.OS + "_" + desc.Arch + ".json"
+}
+
+func (rec *Recorder) Describe() *biome.Descriptor {
+	rec.mu.Lock()
+	defer rec.mu.Unlock()
+	return rec.data.Descriptor
+}
+
+func (rec *Recorder) Dirs() *biome.Dirs {
+	rec.mu.Lock()
+	defer rec.mu.Unlock()
+	return rec.data.Dirs
+}
+
+func (rec *Recorder) Run(ctx context.Context, invoke *biome.Invocation) error {
+	// Intercept I/O and then run in biome.
+	forwarded := &biome.Invocation{
+		Argv: invoke.Argv,
+		Dir:  invoke.Dir,
+		Env:  invoke.Env,
+	}
+	stdinHash := sha256.New()
+	if invoke.Stdin != nil {
+		forwarded.Stdin = io.TeeReader(invoke.Stdin, stdinHash)
+	}
+	var stdout, stderr *bytes.Buffer
+	var combinedOutput *bytes.Buffer
+	if isCombinedOutput(invoke.Stdout, invoke.Stderr) {
+		combinedOutput = new(bytes.Buffer)
+		forwarded.Stdout = io.MultiWriter(invoke.Stdout, combinedOutput)
+		forwarded.Stderr = forwarded.Stdout
+	} else {
+		if invoke.Stdout != nil {
+			stdout = new(bytes.Buffer)
+			forwarded.Stdout = io.MultiWriter(invoke.Stdout, stdout)
+		}
+		if invoke.Stderr != nil {
+			stderr = new(bytes.Buffer)
+			forwarded.Stderr = io.MultiWriter(invoke.Stderr, stderr)
+		}
+	}
```

Comment:
> 1. The Recorder is wrapping the biome, so it's double-writing output; once to the original biome, and once to a recorded location

Correct.

> 2. For ??? reason, it must behave differently if the biome's stdout and stderr are the same location. The way it behaves differently is ???

If stdout and stderr are written to the same place, they must be captured to have them interleave writes correctly. Otherwise, you would have stdout and stderr as wholly separate bytes and there would be no way to sequence them back.

> I think the contributing factors to my difficulty understanding this logic are: [...]

I've dechunked this logic into a few functions that contain the behavior. Sadly, I think it somewhat exacerbates the third issue you outline (because there are more names), but I think the flow is overall better. LMK what you think.